packages/engine/lib/stateful/src/field/spec_map.rs

```diff
@@ -82,14 +89,37 @@ impl<S: FieldSource + Clone> FieldSpecMap<S> {
             if existing_field.scope == FieldScope::Agent
                 && new_field.scope == FieldScope::Agent
                 && existing_field.inner.field_type == new_field.inner.field_type
-                && existing_field.source > new_field.source
             {
-                tracing::warn!(
-                    "Key clash when a package attempted to insert a new agent-scoped field with \
-                     key: {field_key:?}, the existing field was created by a source with a higher \
-                     precedence, the new field will be ignored",
-                );
-                return Ok(());
+                match existing_field.source.partial_cmp(&new_field.source) {
+                    Some(Ordering::Greater) => {
+                        tracing::warn!(
+                            "Key clash when a package attempted to insert a new agent-scoped \
+                             field with key: {field_key:?}, the existing field was created by a \
+                             source with a higher precedence, the new field will be ignored",
+                        );
+                        return Ok(());
+                    }
+                    Some(Ordering::Less) => {
+                        tracing::warn!(
+                            "Key clash when a package attempted to insert a new agent-scoped \
+                             field with key: {field_key:?}, the existing field was created by a \
+                             source with a lower precedence, the existing field will be \
+                             overwritten",
+                        );
+                        self.field_specs.insert(field_key, new_field);
+                        return Ok(());
+                    }
+                    Some(Ordering::Equal) => {
+                        unreachable!("This is expected to be matched before for full equality")
+                    }
+                    None => {
+                        tracing::error!(
```

Comment:
Can we easily say the sources in the error message here? Or is that abstracted away? Like can we end up with a message of "behavior execution" tried adding a field and "messages" already added it

Reply:
Unless we convert `PackageId` to a trait, this isn't really possible.

Reply:
I suspected as much, I don't think it's worth doing right now :) No worries